integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+  _id: BSON.ObjectId;
+  nullableCounter?: Counter | null;
+}
+
+const WithNullableCounterSchema: ObjectSchema = {
+  name: "WithNullableCounter",
+  primaryKey: "_id",
+  properties: {
+    _id: { type: "objectId", default: () => new BSON.ObjectId() },
+    nullableCounter: "counter?",
+  },
+};
+
+interface IWithDefaultCounter {
+  counterWithDefault: Counter;
+}
+
+const WithDefaultCounterSchema: ObjectSchema = {
+  name: "WithDefaultCounter",
+  properties: {
+    counterWithDefault: { type: "int", presentation: "counter", default: 0 },
+  },
+};
+
+interface IWithRegularInt {
+  int: number;
+}
+
+const WithRegularIntSchema: ObjectSchema = {
+  name: "WithRegularInt",
+  properties: {
+    int: "int",
+  },
+};
+
+interface IWithMixed {
+  mixed: Realm.Types.Mixed;
+  list: Realm.List<Realm.Types.Mixed>;
+  dictionary: Realm.Dictionary<Realm.Types.Mixed>;
+  set: Realm.Set<Realm.Types.Mixed>;
+}
+
+const WithMixedSchema: ObjectSchema = {
+  name: "WithMixed",
+  properties: {
+    mixed: "mixed",
+    list: "mixed[]",
+    dictionary: "mixed{}",
+    set: "mixed<>",
+  },
+};
+
+function expectCounter(value: unknown): asserts value is Counter {
+  expect(value).to.be.instanceOf(Counter);
+}
+
+function expectKeys(dictionary: Realm.Dictionary, keys: string[]) {
+  expect(Object.keys(dictionary)).members(keys);
+}
+
+describe("Counter", () => {
+  openRealmBeforeEach({
+    schema: [
+      WithCounterSchema,
+      WithNullableCounterSchema,
+      WithDefaultCounterSchema,
+      WithRegularIntSchema,
+      WithMixedSchema,
+    ],
+  });
+
+  const initialValuesList = [-100, 0, 1.0, 1000] as const;
+
+  describe("Create and access", () => {
+    describe("Via 'realm.create()'", () => {
+      it("can create and access (input: number)", function (this: RealmContext) {
+        for (let i = 0; i < initialValuesList.length; i++) {
+          const input = initialValuesList[i];
+          const { counter } = this.realm.write(() => {
+            return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+              counter: input,
+            });
+          });
+
+          const expectedNumObjects = i + 1;
+          expect(this.realm.objects(WithCounterSchema.name).length).equals(expectedNumObjects);
+          expectCounter(counter);
+          expect(counter.value).equals(input);
+        }
+      });
+
+      it("can create and access (input: Counter)", function (this: RealmContext) {
+        const initialNumValues = initialValuesList;
+        const initialCounterValues: Counter[] = [];
+
+        // First create Realm objects with counters.
+        this.realm.write(() => {
+          for (const input of initialNumValues) {
+            const { counter } = this.realm.create<IWithCounter>(WithCounterSchema.name, {
+              counter: input,
+            });
+            expectCounter(counter);
+            expect(counter.value).equals(input);
+
+            initialCounterValues.push(counter);
+          }
+        });
```

Comment:
A more functional alternative - not feeling strong for this, just a slightly more concise way of expressing this that doesn't require explicit type declaration of `initialCounterValues` nor array manipulation via `push`.

```suggestion
        // First create Realm objects with counters.
        const initialCounterValues = this.realm.write(() => {
          return initialNumValues.map(input => {
            const { counter } = this.realm.create<IWithCounter>(WithCounterSchema.name, {
              counter: input,
            });
            expectCounter(counter);
            expect(counter.value).equals(input);
            return counter;
            });
          }
        });
```